extensions/amp-autocomplete/0.1/amp-autocomplete.js

```diff
@@ -14,51 +14,374 @@
+   * @return {?HTMLElement}
+   * @private
+   */
+  createContainer_() {
+    const container = this.element.ownerDocument.createElement('div');
+    container.classList.add('i-amphtml-autocomplete-results');
+    container.setAttribute('role', 'list');
+    setStyle(container, 'visibility', 'hidden');
+    this.applyFillContent(container, /* replacedContent */ true);
+    return container;
   }
 
   /** @override */
   layoutCallback() {
-    // Actually load your resource or render more expensive resources.
+    this.element.classList.add('i-amphtml-autocomplete');
+    this.inputElement_.classList.add('i-amphtml-autocomplete-input');
+
+    // Disable autofill in browsers.
+    this.inputElement_.setAttribute('autocomplete', 'off');
+
+    // No static local data to filter against.
+    if (!this.inlineData_) {
+      return Promise.resolve();
+    }
+
+    // Register event handlers.
+    this.inputElement_.addEventListener('input',
+        this.inputHandler_.bind(this));
+    this.inputElement_.addEventListener('keydown',
+        this.keyDownHandler_.bind(this));
+    this.inputElement_.addEventListener('focus', this.showResults.bind(this));
+    this.inputElement_.addEventListener('blur', this.hideResults.bind(this));
+    this.renderResults_();
     return Promise.resolve();
   }
 
+  /**
+   * Create and return <div> element from given plan-text item.
+   * @param {string} item
+   * @return {!HTMLElement}
+   * @private
+   */
+  createElementFromItem_(item) {
+    const element = this.element.ownerDocument.createElement('div');
+    element.classList.add('i-amphtml-autocomplete-item');
+    element.setAttribute('role', 'listitem');
+    element.addEventListener('mousedown', this.mousedownHandler_.bind(this));
+    element.textContent = item;
+    return element;
+  }
+
+  /**
+  * Handle rendering results on user input.
+  * @param {?event} event
+  * @private
+  */
+  inputHandler_(event) {
+    if (event.inputType === 'deleteContentBackward') {
+      // Explore options for caching results to avoid repetitive queries.
+    }
+    this.renderResults_();
+    this.showResults();
+  }
+
+  /**
+   * Render filtered results on the current input and update the container_.
+   * @private
+   */
+  renderResults_() {
+    const userInput = this.inputElement_.value;
+    this.clearAllItems();
+    if (userInput.length < this.minChars_) {
+      return;
+    }
+    const filteredData = this.filterData_(this.inlineData_, userInput);
+    filteredData.forEach(item => {
+      this.container_.appendChild(this.createElementFromItem_(item));
+    });
+  }
+
+  /**
+   * Apply the filter to the given data based on the given input.
+   * @param {!Array<string>} data
+   * @param {string} input
+   * @return {!Array<string>}
+   * @private
+   */
+  filterData_(data, input) {
+    let filteredData = data.filter(item => {
+      switch (this.filter_) {
+        case FilterType.SUBSTRING:
+          return item.includes(input);
+          break;
```

Comment:
(Sorry, drive-by.) `break` is not needed after return or throwing. Saves you a few bytes :)

Reply:
🚗